File: src/ui/src-tauri/src/repo.rs
use std::collections::HashMap;

use data::db::infused;
use data::entities::{
    Alert, AlertEvent, App, InteractionPeriod, Period, Reason, Ref, SystemEvent, Tag, Timestamp,
};
use tauri::State;
use util::error::Context;
use util::time::ToTicks;
use util::tracing;

use crate::error::AppResult;
use crate::state::{AppState, Initable, QueryOptions, init_state};

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_apps(
    state: State<'_, AppState>,
    query_options: QueryOptions,
) -> AppResult<HashMap<Ref<App>, infused::App>> {
    let now = query_options.get_now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    let res = repo.get_apps(now).await?;
    Ok(res)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_tags(
    state: State<'_, AppState>,
    query_options: QueryOptions,
) -> AppResult<HashMap<Ref<Tag>, infused::Tag>> {
    let now = query_options.get_now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    let res = repo.get_tags(now).await?;
    Ok(res)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_alerts(
    state: State<'_, AppState>,
    query_options: QueryOptions,
) -> AppResult<HashMap<Ref<Alert>, infused::Alert>> {
    let now = query_options.get_now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    let res = repo.get_alerts(now).await?;
    Ok(res)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_app_durations(
    state: State<'_, AppState>,
    _query_options: QueryOptions,
    start: Timestamp,
    end: Timestamp,
) -> AppResult<HashMap<Ref<App>, infused::WithDuration<App>>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    let res = repo.get_app_durations(start, end).await?;
    Ok(res)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_app_durations_per_period(
    state: State<'_, AppState>,
    _query_options: QueryOptions,
    start: Timestamp,
    end: Timestamp,
    period: Period,
) -> AppResult<HashMap<Ref<App>, Vec<infused::WithGroupedDuration<App>>>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    let res = repo
        .get_app_durations_per_period(start, end, period)
        .await?;
    Ok(res)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_tag_durations_per_period(
    state: State<'_, AppState>,
    _query_options: QueryOptions,
    start: Timestamp,
    end: Timestamp,
    period: Period,
) -> AppResult<HashMap<Ref<Tag>, Vec<infused::WithGroupedDuration<Tag>>>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    let res = repo
        .get_tag_durations_per_period(start, end, period)
        .await?;
    Ok(res)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn copy_from_seed_db(state: State<'_, AppState>) -> AppResult<()> {
    // drop previous state - also drops the db connection
    {
        let mut state = state.write().await;
        state.assume_init().shutdown().await?;
        *state = Initable::Uninit;
    }

    let base_dir = util::config::Config::config_base_dir().context("config base dir")?;
    util::fs::remove_icon_files(&base_dir)?;
    util::fs::remove_db_files(&base_dir)?;

    let to_file = util::config::Config::config_path("main.db").context("config path")?;
    std::fs::copy("../../../dev/seed.db", to_file).context("copy seed.db")?;

    // reinit state for the sake of repo
    init_state(state.clone()).await?;

    {
        let state = state.write().await;
        let mut repo = state.assume_init().get_repo().await?;
        repo.extract_seed_db_icons().await?;
    }

    // shutdown state again, lose the cached type info for sqlx
    {
        let mut state = state.write().await;
        state.assume_init().shutdown().await?;
        *state = Initable::Uninit;
    }
    // reinit state
    init_state(state.clone()).await?;

    Ok(())
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn copy_from_install_db(state: State<'_, AppState>) -> AppResult<()> {
    use util::error::ContextCompat;

    // drop previous state - also drops the db connection
    {
        let mut state = state.write().await;
        state.assume_init().shutdown().await?;
        *state = Initable::Uninit;
    }

    let base_dir = util::config::Config::config_base_dir().context("config base dir")?;
    util::fs::remove_icon_files(&base_dir)?;
    util::fs::remove_db_files(&base_dir)?;

    let install_dir = util::config::data_local_dir()
        .context("data local dir")?
        .join("me.enigmatrix.cobalt");

    util::fs::copy_icon_files(&install_dir, &base_dir)?;
    util::fs::copy_db_files(&install_dir, &base_dir)?;

    // reinit state (repo)
    init_state(state).await?;
    Ok(())
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn update_usages_end(state: State<'_, AppState>) -> AppResult<()> {
    let now = platform::objects::Timestamp::now();
    let state = state.read().await;
    state
        .assume_init()
        .get_repo()
        .await?
        .update_usages_set_last(now.to_ticks())
        .await?;
    Ok(())
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn update_app(state: State<'_, AppState>, app: infused::UpdatedApp) -> AppResult<()> {
    let now = platform::objects::Timestamp::now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    repo.update_app(&app, now).await?;
    Ok(())
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn update_tag(state: State<'_, AppState>, tag: infused::UpdatedTag) -> AppResult<()> {
    let now = platform::objects::Timestamp::now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    repo.update_tag(&tag, now).await?;
    Ok(())
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn update_tag_apps(
    state: State<'_, AppState>,
    tag_id: Ref<Tag>,
    removed_apps: Vec<Ref<App>>,
    added_apps: Vec<Ref<App>>,
) -> AppResult<()> {
    let now = platform::objects::Timestamp::now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    repo.update_tag_apps(tag_id, removed_apps, added_apps, now)
        .await?;
    Ok(())
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn create_tag(
    state: State<'_, AppState>,
    tag: infused::CreateTag,
) -> AppResult<infused::Tag> {
    let now = platform::objects::Timestamp::now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.create_tag(&tag, now).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn remove_tag(state: State<'_, AppState>, tag_id: Ref<Tag>) -> AppResult<()> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.remove_tag(tag_id).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn create_alert(
    state: State<'_, AppState>,
    alert: infused::CreateAlert,
) -> AppResult<Ref<Alert>> {
    let now = platform::objects::Timestamp::now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.create_alert(alert, now).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn update_alert(
    state: State<'_, AppState>,
    prev: infused::Alert,
    next: infused::UpdatedAlert,
) -> AppResult<infused::Alert> {
    let now = platform::objects::Timestamp::now();
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.update_alert(prev, next, now).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn remove_alert(state: State<'_, AppState>, alert_id: Ref<Alert>) -> AppResult<()> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.remove_alert(alert_id).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn create_alert_event(
    state: State<'_, AppState>,
    alert_id: Ref<Alert>,
    timestamp: Timestamp,
    reason: Reason,
) -> AppResult<()> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.create_alert_event(alert_id, timestamp, reason).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_app_session_usages(
    state: State<'_, AppState>,
    start: Timestamp,
    end: Timestamp,
) -> AppResult<infused::AppSessionUsages> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.get_app_session_usages(start, end).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_interaction_periods(
    state: State<'_, AppState>,
    start: Timestamp,
    end: Timestamp,
) -> AppResult<Vec<InteractionPeriod>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.get_interaction_periods(start, end).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_system_events(
    state: State<'_, AppState>,
    start: Timestamp,
    end: Timestamp,
) -> AppResult<Vec<SystemEvent>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.get_system_events(start, end).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_alert_events(
    state: State<'_, AppState>,
    start: Timestamp,
    end: Timestamp,
    alert_id: Ref<Alert>,
) -> AppResult<Vec<AlertEvent>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.get_alert_events(start, end, alert_id).await?)
}

#[tauri::command]
#[tracing::instrument(err, skip(state))]
pub async fn get_alert_reminder_events(
    state: State<'_, AppState>,
    start: Timestamp,
    end: Timestamp,
    alert_id: Ref<Alert>,
) -> AppResult<Vec<infused::ReminderEvent>> {
    let mut repo = {
        let state = state.read().await;
        state.assume_init().get_repo().await?
    };
    Ok(repo.get_alert_reminder_events(start, end, alert_id).await?)
}

File: src/ui/app/lib/repo.ts
import type {
  Alert,
  AlertEvent,
  App,
  Duration as DataDuration,
  InteractionPeriod,
  Ref,
  Reminder,
  ReminderEvent,
  Score,
  Session,
  SystemEvent,
  Tag,
  Target,
  TimeFrame,
  Timestamp,
  TriggerAction,
  WithDuration,
  WithGroupedDuration,
} from "@/lib/entities";
import type { EntityMap, EntityStore } from "@/lib/state";
import { dateTimeToTicks, type Period } from "@/lib/time";
import { invoke } from "@tauri-apps/api/core";
import { DateTime } from "luxon";

export interface CreateTag {
  name: string;
  color: string;
  score: Score;
  apps: Ref<App>[];
}

export type AppSessionUsages = Record<Ref<App>, Record<Ref<Session>, Session>>;

export interface CreateAlert {
  target: Target;
  usageLimit: DataDuration;
  timeFrame: TimeFrame;
  triggerAction: TriggerAction;
  reminders: CreateReminder[];
  ignoreTrigger: boolean;
}

export interface CreateReminder {
  threshold: number;
  message: string;
  ignoreTrigger: boolean;
}

export interface UpdatedAlert {
  id: Ref<Alert>;
  target: Target;
  usageLimit: DataDuration;
  timeFrame: TimeFrame;
  triggerAction: TriggerAction;
  reminders: UpdatedReminder[];
  ignoreTrigger: boolean;
}

export interface UpdatedReminder {
  // undefined when creating a new reminder
  id?: Ref<Reminder>;
  threshold: number;
  message: string;
  ignoreTrigger: boolean;
}

export interface QueryOptions {
  now?: Timestamp;
}

export function getQueryOptions(queryOptions?: QueryOptions): QueryOptions {
  return queryOptions ?? {};
}

export async function getApps({
  options,
}: {
  options?: QueryOptions;
}): Promise<EntityStore<App>> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_apps", { queryOptions });
}

export async function getTags({
  options,
}: {
  options?: QueryOptions;
}): Promise<EntityStore<Tag>> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_tags", { queryOptions });
}

export async function getAlerts({
  options,
}: {
  options?: QueryOptions;
}): Promise<EntityStore<Alert>> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_alerts", { queryOptions });
}

export async function getAppDurations({
  options,
  start,
  end,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
}): Promise<EntityMap<App, WithDuration<App>>> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_app_durations", {
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
    queryOptions,
  });
}

export async function getAppDurationsPerPeriod({
  options,
  start,
  end,
  period,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
  period: Period;
}): Promise<EntityMap<App, WithGroupedDuration<App>[]>> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_app_durations_per_period", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
    period,
  });
}

export async function getTagDurationsPerPeriod({
  options,
  start,
  end,
  period,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
  period: Period;
}): Promise<EntityMap<Tag, WithGroupedDuration<Tag>[]>> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_tag_durations_per_period", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
    period,
  });
}

export async function updateApp(app: App): Promise<void> {
  const updatedApp = {
    id: app.id,
    name: app.name,
    description: app.description,
    company: app.company,
    color: app.color,
    tagId: app.tagId,
  };
  return await invoke("update_app", { app: updatedApp });
}

export async function updateTag(tag: Tag): Promise<void> {
  const updatedTag = {
    id: tag.id,
    name: tag.name,
    color: tag.color,
    score: tag.score,
  };
  return await invoke("update_tag", { tag: updatedTag });
}

export async function updateTagApps(
  tagId: Ref<Tag>,
  removedApps: Ref<App>[],
  addedApps: Ref<App>[],
): Promise<void> {
  return await invoke("update_tag_apps", { tagId, removedApps, addedApps });
}

export async function createTag(tag: CreateTag): Promise<Tag> {
  return await invoke("create_tag", { tag });
}

export async function removeTag(tagId: Ref<Tag>): Promise<void> {
  return await invoke("remove_tag", { tagId });
}

export async function createAlert(alert: CreateAlert): Promise<Alert> {
  return await invoke("create_alert", { alert });
}

export async function updateAlert(
  prev: Alert,
  next: UpdatedAlert,
): Promise<Alert> {
  return await invoke("update_alert", { prev, next });
}

export async function removeAlert(alertId: Ref<Alert>): Promise<void> {
  return await invoke("remove_alert", { alertId });
}

export async function createAlertEventIgnore(
  alertId: Ref<Alert>,
  timestamp: Timestamp,
): Promise<void> {
  return await invoke("create_alert_event_ignore", { alertId, timestamp });
}

export async function getAppSessionUsages({
  options,
  start,
  end,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
}): Promise<AppSessionUsages> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_app_session_usages", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
  });
}

export async function getInteractionPeriods({
  options,
  start,
  end,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
}): Promise<InteractionPeriod[]> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_interaction_periods", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
  });
}

export async function getSystemEvents({
  options,
  start,
  end,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
}): Promise<SystemEvent[]> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_system_events", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
  });
}

export async function getAlertEvents({
  options,
  start,
  end,
  alertId,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
  alertId: Ref<Alert>;
}): Promise<AlertEvent[]> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_alert_events", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
    alertId,
  });
}

export async function getAlertReminderEvents({
  options,
  start,
  end,
  alertId,
}: {
  options?: QueryOptions;
  start: DateTime;
  end: DateTime;
  alertId: Ref<Alert>;
}): Promise<ReminderEvent[]> {
  const queryOptions = getQueryOptions(options);
  return await invoke("get_alert_reminder_events", {
    queryOptions,
    start: dateTimeToTicks(start),
    end: dateTimeToTicks(end),
    alertId,
  });
}
